src/OpenTelemetry/Trace/SpanSdk.cs

```diff
@@ -620,7 +620,7 @@ public void Dispose()
             {
                 spanCreationOptions = new SpanCreationOptions
                 {
-                    Attributes = activity.Tags.ToDictionary(k => k.Key, v => (object)v.Value),
+                    Attributes = activity.Tags.Cast<KeyValuePair<string, object>>(),
```

Comment:
Attributes in SpanCreationOptions are `KeyValuePair<string,object>`, while Activity.Tags are `KeyValuePair<string,string>`. So we cannot assign the tags to attributes.

@tarekgh Is there a way to assign this without allocating a new object?

Reply:
You are changing the Attributes to be IEnumerable<KeyValuePair<string, object>> so a simple cast should just do it without any allocation.

```C#
Attributes = (IEnumerable<KeyValuePair<string, object>>) activity.Tags;
```

Reply:
The simple cast above is throwing exception at runtime during the cast.
the Tags.Cast<KeyValuePair<string, object>>, not throwing at the point, but when iterating over it.

```
            Activity a = new Activity("cijo");
            a.AddTag("key", "value1");
            IEnumerable<KeyValuePair<string, object>> col1;            
            col1 = (IEnumerable<KeyValuePair<string, object>>) a.Tags;  // throws here at runtime
```
```
System.InvalidCastException: 'Unable to cast object of type '<<get_Tags>g__Iterate|44_0>d' to type 'System.Collections.Generic.IEnumerable`1[System.Collections.Generic.KeyValuePair`2[System.String,System.Object]]'.'
```


```
            Activity a = new Activity("cijo");
            a.AddTag("key", "value1");
            IEnumerable<KeyValuePair<string, object>> col1;                        
            col1 = a.Tags.Cast<KeyValuePair<string, object>>();  // doesnt throw here

            foreach (var attribute in col1)  // throws here
            {

            }
```
System.InvalidCastException: 'Specified cast is not valid.'
```

```

Reply:
I'll take a look and get back to you.

Reply:
please try 

```C#
    internal struct TagsCollection : IEnumerable<KeyValuePair<string, object>>
    {
        private IEnumerable<KeyValuePair<string, string>> _enuemrable;
        public TagsCollection(IEnumerable<KeyValuePair<string, string>> tags) { _enuemrable = tags; }
        public IEnumerator GetEnumerator() => new TagsEnumerator(_enuemrable.GetEnumerator());
        IEnumerator<KeyValuePair<string, object>> IEnumerable<KeyValuePair<string, object>>.GetEnumerator() => new TagsEnumerator(_enuemrable.GetEnumerator());

        private struct TagsEnumerator : IDisposable, IEnumerator<KeyValuePair<string, object>>
        {
            IEnumerator<KeyValuePair<string, string>> _enumerator;
            public TagsEnumerator(IEnumerator<KeyValuePair<string, string>> enumerator) => _enumerator = enumerator;
            KeyValuePair<string, object> IEnumerator<KeyValuePair<string, object>>.Current => new KeyValuePair<string, object>(_enumerator.Current.Key, (object) _enumerator.Current.Value);
            object IEnumerator.Current => new KeyValuePair<string, object>(_enumerator.Current.Key, (object) _enumerator.Current.Value);
            public void Dispose() => _enumerator.Dispose();
            public bool MoveNext() => _enumerator.MoveNext();
            public void Reset() => _enumerator.Reset();
        }
    }
```

and then you can do:

```C#
Attributes = new TagsCollection(activity.Tags);
```

Let me know if you see any problem with it.